cmd/charm/charmcmd/cmd.go

```diff
@@ -201,3 +202,42 @@ func validateAuthFlag(flagval string) (string, string, error) {
 func ussoTokenPath() string {
 	return osenv.JujuXDGDataHomePath("store-usso-token")
 }
+
+// errorMessage translates err into a user understandable error message,
+// and outputs the message ctxt.Stderr. If a message is output then
+// cmd.ErrSilent is returned, othewise err is returned unchanged.
+func errorMessage(ctxt *cmd.Context, err error) error {
+	if err == nil {
+		return err
+	}
+	cause := errgo.Cause(err)
+	switch {
+	case httpbakery.IsInteractionError(cause):
+		return interactionErrorMessage(ctxt, cause.(*httpbakery.InteractionError))
+	}
+	return err
+}
+
+// interactionErrorMessage translates err into a user understandable error message,
+// and outputs the message ctxt.Stderr.
+func interactionErrorMessage(ctxt *cmd.Context, err *httpbakery.InteractionError) error {
+	reason := err.Reason.Error()
+	if ussoError, ok := errgo.Cause(err.Reason).(*usso.Error); ok {
+		reason = ussoError.Error()
+		if ussoError.Code == "INVALID_DATA" && len(ussoError.Extra) > 0 {
+			for k, v := range ussoError.Extra {
```

Comment:
Perhaps comment that you expect only one entry in the Extra map.
If not, you're going to choose an arbitrary error message here AFAICS.